### ISSUE TITLE
package-lock.json doesn't produce the expected behaviour

### ISSUE THREAD

**Issue by hi-larry** [npm/npm#20660]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [x] npm is producing incorrect or undesirable behavior.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
Using Renovate (@rarkins), I noticed that package-lock.json doesn't produce the same results upon updating the version of a package manually (and running `npm install --package-lock-only`) and updating via CLI using the command `npm install foo@version --package-lock-only`. This cause a significant problem as the dependencies (that have the same package to update) do not have the same version as the main.

I made an example to reproduce this behaviour. Notice the difference between [this](https://github.com/hi-larry/TestDep/blob/manually/LCS/package-lock.json#L7-L10) and [that](https://github.com/hi-larry/TestDep/blob/command-only/LCS/package-lock.json#L7-L16). The second file is the expected behaviour while the first one is missing the dependency that is being updated.

#### How can the CLI team reproduce the problem?
Please visit [this example](https://github.com/hi-larry/TestDep).

### supporting information:

 - `npm -v` prints: v6.0.1
 - `node -v` prints: v8.9.0
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: macOS
 - Network issues:
   - Geographic location where npm was run:
   - [x] I use a proxy to connect to the npm registry.
   - [x] I use a proxy to connect to the web.
   - [x] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by legodude17** [comment]
The reason the command made a different one is because the command doesn't change the version of `ioredis` that `bar` requires, but you did manually. So the command install `ioredis@3.2.2` in your top level, and then keeps `ioredis@2.5.0` inside `bar`. But you changed all of them so `ioredis@3.2.2` is deduped to the top.

**Comment by hi-larry** [comment]
Does that mean you don't recommend manually changing any versions in package.json as it won't generate the same package-lock.json as installing a different version via CLI?

**Comment by legodude17** [comment]
Wait I thought you edited `package-lock.json`. What file did you edit? Also, did you ever touch `bar`?

**Comment by hi-larry** [comment]
No, I edited `package.json`. I never manually edited `package-lock.json`. And no, didn't change `bar` either. In my repo, the "package" LCS has `foo` and `bar` as its dependencies, and `ioredis` as its devDependencies. However, `foo` and `bar` also have `ioredis` as their dependency but one with a newer version and another with an older version. This experiment is used to see if running the command would update the `package-lock.json` accordingly.

**Comment by legodude17** [comment]
Well, `npm install --package-lock-only` would completely ignore your change in `package.json`. But it doesn't appear that it did. I'm confused now

**Comment by rarkins** [comment]
Why would npm install —package-lock-only ignore changes to package.json? I think that’s not a correct statement. The command should produce a valid lock file. If it does not, then it’s a bug.

________________________________

**Comment by rarkins** [comment]
Smaller way to reproduce the problem:

- `package.json` has dependencies on `package-a` and `package-b@^1.0.0`
- `package-a` also has a dependency on `package-b@^1.0.0`
- now update `package.json` to depend on `package-b@^2.0.0`
- The `package-lock.json` generated is then invalid - it drops `package-a`'s transitive dependency on `package-b`, i.e. the 	requires: { package-b: "^1.0.0" }` section disappears
- it only works if you delete both `node_modules` and `package-lock.json` and use `npm i` instead of `npm i --package-lock-only`
- it also works if - instead of manually patching the version in `package.json` - you run `npm i package-b@^2.0.0 `with or without `--package-lock-only`

**Comment by hi-larry** [comment]
Now that the issue is more clear, would it be possible for Renovate to delete `node_modules` and `package-lock.json` in order to generate the right result of `package-lock.json`?

**Comment by hi-larry** [comment]
@legodude17 What would be the recommended way to retrieve the right `package-lock.json` ? As @rarkins mentioned [here](https://github.com/renovateapp/renovate/issues/1984#issuecomment-389983275), it would generate a faulty `package-lock.json` if we don't delete `node_modules` or `package-lock.json` beforehand.

**Comment by rarkins** [comment]
It looks like there was a related fix for this in npm@6.1.0 however I am still not seeing correct results with the example I'm using (adapted off @hi-larry's example).

To reproduce:
```
npm i -g npm@6.1.0
git clone https://github.com/rarkins/package-lock-update
npm i --package-lock-only
git diff | less
```

You can see that the transitive `ioredis` requires is still being removed:

```diff
diff --git a/package-lock.json b/package-lock.json
index a22925a..bd9bb1d 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -6,73 +6,185 @@
   "dependencies": {
     "bar": {
       "version": "file:bar-1.0.0.tgz",
-      "integrity": "sha512-i0C/TLZdUgjs7IUGebzGMtmHD2OI63mrGNCT2Ck5fIerzvQrdmsIf0uCS2oMEn9+Vc7rpRADomeLwyx7YzcNrQ==",
-      "requires": {
-        "ioredis": "^2.5.0"
-      }
+      "integrity": "sha512-i0C/TLZdUgjs7IUGebzGMtmHD2OI63mrGNCT2Ck5fIerzvQrdmsIf0uCS2oMEn9+Vc7rpRADomeLwyx7YzcNrQ=="
     },
```

This alternative where the old lock file is removed first does work though:
```
npm i -g npm@6.1.0
git clone https://github.com/rarkins/package-lock-update
rm -f package-lock.json
npm i --package-lock-only
git diff | less
```

Result:
```diff
diff --git a/package-lock.json b/package-lock.json
index a22925a..0e730a2 100644
--- a/package-lock.json
+++ b/package-lock.json
@@ -9,6 +9,23 @@
       "integrity": "sha512-i0C/TLZdUgjs7IUGebzGMtmHD2OI63mrGNCT2Ck5fIerzvQrdmsIf0uCS2oMEn9+Vc7rpRADomeLwyx7YzcNrQ==",
       "requires": {
         "ioredis": "^2.5.0"
+      },
+      "dependencies": {
+        "ioredis": {
+          "version": "2.5.0",
+          "resolved": "https://registry.npmjs.org/ioredis/-/ioredis-2.5.0.tgz",
+          "integrity": "sha1-+2/fChp+CXRhTGe25eETCKjPlbk=",
+          "requires": {
+            "bluebird": "^3.3.4",
+            "cluster-key-slot": "^1.0.6",
+            "debug": "^2.2.0",
+            "double-ended-queue": "^2.1.0-0",
+            "flexbuffer": "0.0.6",
+            "lodash": "^4.8.2",
+            "redis-commands": "^1.2.0",
+            "redis-parser": "^1.3.0"
+          }
+        }
       }
     },
```
